### DISCUSSION TITLE
Conv type is missing

### DISCUSSION THREAD

**Discussion by rbcarlos** [Xilinx/finn#461] [Q&A]
The brevitas export to onnx throws following warnings:
```
WARNING: The shape inference of finn.custom_op.general::Conv type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
```

It does so for MatMul, Conv and Multithreshold. 
This causes problems as the types default to floats32 and therefore the to_hls transforms cannot continue as they expect ints.

Here are examples of definitions of all the layers.
```
brevitas.nn.QuantConv1d(in_channels = hid_dim, 
                            out_channels = hid_dim, 
                            kernel_size = kernel_size, 
                            padding = (kernel_size - 1) // 2,
                            #input_quant=Int8ActPerTensorFloat,
                            return_quant_tensor=True,
                            bias=False,
                            weight_quant=CommonWeightQuant,
                weight_bit_width=weight_bit_width)

brevitas.nn.QuantLinear(emb_dim, hid_dim, bias=False,
                weight_bit_width=weight_bit_width,
                return_quant_tensor=True,
                weight_quant=CommonWeightQuant)

brevitas.nn.QuantIdentity(
                bit_width=weight_bit_width, 
                return_quant_tensor=True, 
                act_quant=CommonActQuant,
                min_val=- 1.0,
                max_val=1.0 - 2.0 ** (-7),
                narrow_range=False,
                restrict_scaling_type=RestrictValueType.POWER_OF_TWO
            )
```

When I try to explicitly set the input_quant the brevitas export throws an error.

**Comment by maltanar** [comment]
Thanks for sharing this @rbcarlos , this looks like a potential Brevitas bug. `QuantConv` and `QuantLinear` should be exporting regular ONNX ops without any custom domain string. Could you share which version of Brevitas you are using, and how you are doing the export to ONNX?

**Reply by rbcarlos** [reply]
We are running FINN v0.7 in docker, so BREVITAS_COMMIT="a5b71d6de1389d3e7db898fef72e014842670f03". 

I have actually found out that it has nothing to do with the warnings during the export (I get these also with other exports that run through FINN smoothly), but rather with the network defaulting to float32 datatypes in the outputs of some of the layers (mostly mult by a scalar), which prevents the conversion to hls from happening. Is there any way to specify the datatypes manually when this happens? 

Regarding the export we used regular:
```
bo.export_finn_onnx(model.model.to("cpu"), (1, 200, 50), "classifier_1.onnx")
```
